File: lib/decoders/fmmfm.cc
#include "lib/core/globals.h"
#include "lib/decoders/decoders.h"

Bytes decodeFmMfm(
    std::vector<bool>::const_iterator ii, std::vector<bool>::const_iterator end)
{
    /*
     * FM is dumb as rocks, consisting on regular clock pulses with data pulses
     * in the gaps. 0x00 is:
     *
     *     X-X-X-X-X-X-X-X-
     *
     * 0xff is:
     *
     *     XXXXXXXXXXXXXXXX
     *
     * So we just need to extract all the odd bits.
     *
     * MFM and M2FM are slightly more complicated, where the first bit of each
     * pair can be either 0 or 1... but the second bit is always the data bit,
     * and at this point we simply don't care what the first bit is, so
     * decoding MFM uses just the same code!
     */

    Bytes bytes;
    ByteWriter bw(bytes);

    int bitcount = 0;
    uint8_t fifo = 0;

    while (ii != end)
    {
        ii++; /* skip clock bit */
        if (ii == end)
            break;
        fifo = (fifo << 1) | *ii++;

        bitcount++;
        if (bitcount == 8)
        {
            bw.write_8(fifo);
            bitcount = 0;
        }
    }

    if (bitcount != 0)
    {
        fifo <<= 8 - bitcount;
        bw.write_8(fifo);
    }

    return bytes;
}

void encodeFm(std::vector<bool>& bits, unsigned& cursor, const Bytes& input)
{
    if (bits.size() == 0)
        return;
    unsigned len = bits.size() - 1;

    for (uint8_t b : input)
    {
        for (int i = 0; i < 8; i++)
        {
            bool bit = b & 0x80;
            b <<= 1;

            if (cursor >= len)
                return;

            bits[cursor++] = true;
            bits[cursor++] = bit;
        }
    }
}

void encodeMfm(std::vector<bool>& bits,
    unsigned& cursor,
    const Bytes& input,
    bool& lastBit)
{
    if (bits.size() == 0)
        return;
    unsigned len = bits.size() - 1;

    for (uint8_t b : input)
    {
        for (int i = 0; i < 8; i++)
        {
            bool bit = b & 0x80;
            b <<= 1;

            if (cursor >= len)
                return;

            bits[cursor++] = !lastBit && !bit;
            bits[cursor++] = bit;
            lastBit = bit;
        }
    }
}

Bytes encodeMfm(const Bytes& input, bool& lastBit)
{
    ByteReader br(input);
    BitReader bitr(br);
    Bytes b;
    ByteWriter bw(b);
    BitWriter bitw(bw);

    while (!bitr.eof())
    {
        uint8_t bit = bitr.get();

        bitw.push(!lastBit && !bit);
        bitw.push(bit);
        lastBit = bit;
    }

    bitw.flush();
    return b;
}

File: java/com/cowlark/fluxengine/external/FmMfm.java
package com.cowlark.fluxengine.external;

import com.cowlark.fluxengine.core.BitReader;
import com.cowlark.fluxengine.core.BitWriter;
import com.cowlark.fluxengine.core.Bits;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.ByteWriter;

/**
 * FM and MFM encode/decode helpers, ported from lib/decoders/fmmfm.cc.
 * The {@code cursor} and {@code lastBit} parameters of the encoding functions
 * are carried in single-element arrays to provide the in/out semantics of the
 * C++ references.
 */
public final class FmMfm
{
    private FmMfm()
    {
    }

    /*
     * FM is dumb as rocks, consisting on regular clock pulses with data pulses
     * in the gaps. 0x00 is:
     *
     *     X-X-X-X-X-X-X-X-
     *
     * 0xff is:
     *
     *     XXXXXXXXXXXXXXXX
     *
     * So we just need to extract all the odd bits.
     *
     * MFM and M2FM are slightly more complicated, where the first bit of each
     * pair can be either 0 or 1... but the second bit is always the data bit,
     * and at this point we simply don't care what the first bit is, so
     * decoding MFM uses just the same code!
     */
    public static Bytes decodeFmMfm(Bits bits)
    {
        Bytes bytes = new Bytes(0);
        ByteWriter bw = new ByteWriter(bytes);

        int bitcount = 0;
        int fifo = 0;
        int i = 0;
        while (i < bits.size())
        {
            i++; /* skip clock bit */
            if (i >= bits.size())
                break;
            fifo = (fifo << 1) | (bits.getBit(i++) ? 1 : 0);

            bitcount++;
            if (bitcount == 8)
            {
                bw.write8(fifo);
                bitcount = 0;
            }
        }

        if (bitcount != 0)
        {
            fifo <<= 8 - bitcount;
            bw.write8(fifo);
        }

        return bytes;
    }

    public static void encodeFm(Bits bits, int[] cursor, Bytes input)
    {
        if (bits.size() == 0)
            return;
        int len = bits.size() - 1;

        for (int i = 0; i < input.size(); i++)
        {
            int b = input.getByte(i) & 0xff;
            for (int j = 0; j < 8; j++)
            {
                boolean bit = (b & 0x80) != 0;
                b <<= 1;

                if (cursor[0] >= len)
                    return;

                bits.set(cursor[0]++, true);
                bits.set(cursor[0]++, bit);
            }
        }
    }

    public static void encodeMfm(
        Bits bits, int[] cursor, Bytes data, boolean[] lastBit)
    {
        if (bits.size() == 0)
            return;
        int len = bits.size() - 1;

        for (int i = 0; i < data.size(); i++)
        {
            int b = data.getByte(i) & 0xff;
            for (int j = 0; j < 8; j++)
            {
                boolean bit = (b & 0x80) != 0;
                b <<= 1;

                if (cursor[0] >= len)
                    return;

                bits.set(cursor[0]++, !lastBit[0] && !bit);
                bits.set(cursor[0]++, bit);
                lastBit[0] = bit;
            }
        }
    }

    public static Bytes encodeMfm(Bytes data, boolean[] lastBit)
    {
        ByteReader br = new ByteReader(data);
        BitReader bitr = new BitReader(br);
        Bytes out = new Bytes(0);
        ByteWriter bw = new ByteWriter(out);
        BitWriter bitw = new BitWriter(bw);

        while (bitr.hasNext())
        {
            boolean bit = bitr.next();

            bitw.push(!lastBit[0] && !bit);
            bitw.push(bit);
            lastBit[0] = bit;
        }

        bitw.flush();
        return out;
    }
}